testing-guidelines/ci.md

```diff
@@ -0,0 +1,431 @@
+> It isn't a big deal but we need to fix them ASAP and try to make sure it doesn't happen again in the future.
+
+These changesets appear in the `.changeset` directory as markdown files and each changeset has its own dedicated file (this reduces conflicts). These changeset can be written manually or generated via a CLI but we try to automatically detect the changes in a PR and generate a changeset. This isn't always perfect but it is pretty reliable. The changeset can be updated by modifying labels, editing the issue, or modifying the file directly.
+
+#### merging into `main`
+
+When a pull request is merging to main there are two possible routes.
+
+- Create or update a versioning PR
+- Publish packages
+
+A normal pull request created by a user that does not contain a changeset will do nothing.
+
+A normal pull request created by a user that _does_ contain a changeset will either create or update a versioning PR. This is a PR that tracks the changelog and versions that will be used for the release. The changelogs for each package are generated from the descriptions in the changeset files. The generated version pull request is on a branch named `changeset-release/main` and there is only ever one (unless we are also working on a pre-release).
+
+When the `changeset-release/main` branch is merged into main all of the necessary versions will have been bumped and the changelogs generated. We then go through to see what needs to be published and publish to PyPi for Python packages and npm for javascript pacakges.
+
+
+## Architecture
+
+The CI on this repo is a little unconventional, this is mainly to work around various gaps in the GitHub Actions API while solving for our use case.
+
+The [technical details](#technical-details) below explain how things are setup, the [context](#context) sections explains what problem we are trying to solve with this architecture, why we took this approach, and the various challenges we faced along the way. It is a little 'extra' for repo docs but serves as good context for the future.
+
+### Technical details
+
+Most our our workflows are contained within the repository but we also have some custom JavaScript actions that are stored in the [gradio-app/github](https://github.com/gradio-app/github) repository. These actions will be mentioned below but I'll put a list here for ease of navigation:
+
+- [`find-pr`](https://github.com/gradio-app/github/tree/main/packages/find-pr) - for getting a pr number, pr branch, and other metadata for a `workflow_run`.
+- [`comment-pr`](https://github.com/gradio-app/github/tree/main/packages/comment-pr) - for creating and updating comments on a pull request.
+- [`commit-status`](https://github.com/gradio-app/github/tree/main/packages/commit-status) - for creating statuses on commits.
+- [`generate-changeset`](https://github.com/gradio-app/github/tree/main/packages/generate-changeset) - for generating and updating changesets based on the state or a pr and branch.
```

Comment:
TBH, I've been wondering (for my own projects and so on, but the same wonderment applies here) why not write this stuff into a couple Python scripts instead...

Reply:
I guess it depends, we publish a lot of packages and want the process to be very simple, this allows us to write less code ourselves. There is no reason why a few scripts couldn't replace this for a specific case.

That said, I do have vague plans to pull this out and package it up in way that requires less custom code and just some simple config.